File: advent of code/2021/java/day7.java
/**
 * Advent of Code 2021 - Day 7: The Treachery of Whales
 * 6 December 2021
 * 
 * part 1: 
 * part 2: 
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class day7 {

    public static void main(String[] args) throws FileNotFoundException {
        part1();
        part2();
    }

    public static void part1() throws FileNotFoundException {
        File input = new File("inputs/day7.txt");

        Scanner sc = new Scanner(input);
        String[] inputs = sc.nextLine().split(",");

        ArrayList<Integer> crabPosition = new ArrayList<>();

            for(String in : inputs) {
            crabPosition.add(Integer.parseInt(in));
        }

        double ave = 0;

        for(int i = 0; i < crabPosition.size(); i++) {
            ave += crabPosition.get(i);
        }

        int average = Math.round((float)ave /crabPosition.size());

        System.out.println(average);

        // finding 20% around the average

        int lowerBound = (int) Math.round((float)average * 0.3);
        int upperBound = (int) Math.round((float)average * 0.2) + average;

        System.out.println(lowerBound + " " + upperBound);

        double leastFuel = Double.POSITIVE_INFINITY;

        for(int i = lowerBound; i < upperBound; i++) {
            int currFuel = 0;
            
            for(int crab : crabPosition) {
                currFuel += Math.abs(crab - i);
            }

            if(currFuel < leastFuel) leastFuel = currFuel;
        }

        System.out.println(leastFuel);

        sc.close();
    }

    public static void part2() throws FileNotFoundException {
        File input = new File("inputs/day7.txt");

        Scanner sc = new Scanner(input);
        String[] inputs = sc.nextLine().split(",");

        ArrayList<Integer> crabPosition = new ArrayList<>();

        for(String in : inputs) {
            crabPosition.add(Integer.parseInt(in));
        }

        double ave = 0;

        for(int i = 0; i < crabPosition.size(); i++) {
            ave += crabPosition.get(i);
        }

        int average = Math.round((float)ave /crabPosition.size());

        System.out.println(average);

        // finding 20% around the average

        int lowerBound = (int) Math.round((float)average * 0.3);
        int upperBound = (int) Math.round((float)average * 0.2) + average;

        System.out.println(lowerBound + " " + upperBound);

        long leastFuel = 100000000;

        for(int i = lowerBound; i < upperBound; i++) {
            int currFuel = 0;
            
            for(int crab : crabPosition) {

                for(int step = 1; step <= Math.abs(crab - i); step++) {
                    currFuel += step;
                }
            }

            if(currFuel < leastFuel) leastFuel = currFuel;
        }

        System.out.println(leastFuel);

        sc.close();
    }
}

File: advent of code/2021/python/day7.py
# Advent of Code 2021 - Day 7: The Treachery of Whales
# 7 December 2021
# 
# 20 minutes

def limitToMiddleThird(input):
    sum = 0
    for line in input:
        sum += int(line) # assuming input is a string of numbers
    
    average = sum / len(input)

    lowerBound = round(average * 0.33)
    upperBound = round(average) + round(average * 0.165)

    return [lowerBound, upperBound]

# reading input
f = open('inputs/day7.txt', 'r')
lines = f.readlines()

lines = lines.pop().rstrip().split(",")

# part 1
finSum = float('inf')
totalSum = 0

bounds = limitToMiddleThird(lines)

for i in range(int(bounds[0]), int(bounds[1])):
    for line in lines:
        totalSum += abs(i-int(line))
    
    if(totalSum < finSum):
        finSum = totalSum
    
    totalSum = 0

print(finSum)

# part 2
finSum = float('inf')
totalSum = 0

for i in range(int(bounds[0]), int(bounds[1])):
    for line in lines:

        for step in range(1, abs(i-int(line))+1 ): # slow for some reason
            totalSum += step
    
    if(totalSum < finSum):
        finSum = totalSum
    
    totalSum = 0

print(finSum)